[torchgen] Generate wrapper functions under custom namespaces

larryliu0820 opened the pull request:
Summary:
A follow up of #81581. Before these 2 PRs, if an operator with custom kernel namespace is added to `native_functions.yaml` (or any other yaml consumed by `torchgen`), although we are able to recognize the custom kernel in files such as `NativeFunctions.h` and `RegisterCPU.cpp`, we still generate backend specific wrappers under the hardcoded `at` namespace. This changes the behavior, by generating wrapper functions under custom namespaces.

For example, if the entries in yaml file looks like:

```
 - func: op_1(Tensor(a) self) -> Tensor(a)
  dispatch:
    CPU: at::op_1_kernel # ATen kernel

- func: op_2(Tensor(a) self) -> Tensor(a)
  dispatch:
    CPU: custom::op_2_kernel # custom kernel
```

We generate the following code for `CPUFunctions_inl.h` and `RegisterCPU.cpp`:

`CPUFunctions_inl.h`:
```
namespace at {
namespace cpu {
TORCH_API at::Tensor & op_1(const at::Tensor & self);
} // namespace cpu
} // namespace at

namespace custom {
namespace cpu {
TORCH_API at::Tensor & op_2(const at::Tensor & self);
} // namespace cpu
} // namespace custom

```

Notice the difference between `at::cpu` and `custom::cpu`.

Then the definition for these can be found in `RegisterCPU.cpp`.

`RegisterCPU.cpp`:
```
#include "CPUFunctions.h"

namespace at {

namespace {
at::Tensor & wrapper_op_1(const at::Tensor & self) {
    // No device check
  // DeviceGuard omitted
  return at::native::op_1_kernel(self);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {
m.impl("op_1", TORCH_FN(wrapper_op_1));
}

namespace cpu {
at::Tensor & op_1(at::Tensor & self) {
  return wrapper_op_1(self);
}
} // namespace cpu
} // namespace at

namespace custom {

namespace {
at::Tensor & wrapper_op_2(const at::Tensor & self) {
    // No device check
  // DeviceGuard omitted
  return at::native::op_2_kernel(self);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {
m.impl("op_2", TORCH_FN(wrapper_op_2));
}

namespace cpu {
at::Tensor & op_2(at::Tensor & self) {
  return wrapper_op_2(self);
}
} // namespace cpu
} // namespace custom

```

The benefit for this change is that it unifies all the namespaces derived from custom ops. In the example above, there are:

1. `custom::native` for kernels
2. `custom::<dispatch_key>` e.g., `custom::cpu` for wrappers

This customized operator will have nothing to do with `at::native`, `at::cpu` etc.

Test Plan: This is very hard to test. I will refactor this logic, abstract out some layers so it's testable. Will do it in coming PRs

Differential Revision: D37972772

ezyang commented:
I'm going to let Brian take this.

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1399,6 +1399,165 @@ def get_native_function_declarations(
+            backend_metadata.cpp_namespace
+            if backend_metadata
+            else DEFAULT_KERNEL_NAMESPACE
+        ).replace("::native", "")
```
what's going on here? (I know that on the mobile side we'd rather not have the `native` bit show up in the namespace, but how does this work on the core side?)

I think it would be cleaner to arrange for the correct namespace to already be set as a field on the `NativeFunction` or `BackendMetadata` somewhere, so you don't need to do last-minute string parsing.

larryliu0820 replied:
it's just for the `at::<dispatch_key>` namespace, it's declared in `<DispatchKey>Functions_inl.h`. Example: https://fburl.com/67fm7say

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1399,6 +1399,165 @@ def get_native_function_declarations(
+
+}} // anonymous namespace
+
+namespace {dispatch_key.lower()} {{
```
so, do we have to worry about multiple namespaces showing up the same file? (or is it always "aten plus one more namespace").

I think I find this whole function a bit harder to reason about than the previous code - before I could look at `RegisterDispatchKey.cpp and see the 3 distinct pieces of logic (anonymous definitions, namespaced definitions, registrations), as well as what namespace they were in (hardcoded to ATen).

In this PR we still want all of the same stuff, except for another namespace to be there. Can we keep more of the information directly in the C++ template file? E.g:
```
namespace ${op_namespace} {
namespace {
${dispatch_helpers}
${dispatch_anonymous_definitions}
${static_init_dispatch_registrations}
} // anonymous namespace
${deferred_dispatch_registrations}
namespace ${dispatch_namespace} {
${dispatch_namespaced_definitions}
} // namespace ${dispatch_namespace}
} // namespace ${op_namespace}
```
(or if "ATen" + "custom namespace" need to be able to coexist in the same file, then we could just stamp out two of these in the template?)

swolchok replied:
I'm here from the future trying to implement #143952 and I don't understand why we had to create another layer of templating in RegisterDispatchDefinitions.ini rather than embedding the namespace directly in RegisterDispatchKey.cpp. Can anybody shed some light on this?

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1403,6 +1404,160 @@ def get_native_function_declarations(
+            backend_metadata.cpp_namespace
+            if backend_metadata
+            else DEFAULT_KERNEL_NAMESPACE
+        ).replace("native", dispatch_key.lower())
```
Is `namespace` here guaranteed to end in `native`? For the core codegen, I see that `DEFAULT_KERNEL_NAMESPACE` is `at::native`, so this will correctly replace the namespace with e.g. `at::cpu`. Is the idea that for executorch, we'd expect `cpp_namespace here to be something like `executorch::native`?, and so these kernels will live in `executorch::cpu`?

If it is, we should probably assert that the backend namespace's inner-most namespace is always `native`.

larryliu0820 replied:
We append `native` keyword unconditionally to the namespace read from yaml. That logic is in [model.py](https://github.com/pytorch/pytorch/blob/master/torchgen/model.py#L636-L649). But good point I'll add assert here.

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1403,6 +1404,160 @@ def get_native_function_declarations(
+            backend_metadata.cpp_namespace
+            if backend_metadata
+            else DEFAULT_KERNEL_NAMESPACE
+        ).replace("::native", "")
```
Same thing here (is the backend namespace guaranteed to end in `native`? maybe assert it and add a comment explaining why)

bdhirsh reviewed (approved):
left some light comments, but preemptively approving!

larryliu0820 committed:
[torchgen] Generate wrapper functions under custom namespaces (#81744)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81744

Before this change, if an operator with custom kernel namespace is added to `native_functions.yaml` (or any other yaml consumed by `torchgen`), although we are able to recognize the custom kernel in files such as `NativeFunctions.h` and `RegisterCPU.cpp`, we still generate backend specific wrappers under the hardcoded `at` namespace. This changes the behavior, by generating wrapper functions under custom namespaces.

For example, if the entries in yaml file looks like:

```
 - func: op_1(Tensor(a) self) -> Tensor(a)
  dispatch:
    CPU: at::op_1_kernel # ATen kernel

- func: op_2(Tensor(a) self) -> Tensor(a)
  dispatch:
    CPU: custom::op_2_kernel # custom kernel
```

We generate the following code for `CPUFunctions_inl.h` and `RegisterCPU.cpp`:

`CPUFunctions_inl.h`:
```
namespace at {
namespace cpu {
TORCH_API at::Tensor & op_1(const at::Tensor & self);
} // namespace cpu
} // namespace at

namespace custom {
namespace cpu {
TORCH_API at::Tensor & op_2(const at::Tensor & self);
} // namespace cpu
} // namespace custom

```

Notice the difference between `at::cpu` and `custom::cpu`.

Then the definition for these can be found in `RegisterCPU.cpp`.

`RegisterCPU.cpp`:
```
#include "CPUFunctions.h"

namespace at {

namespace {
at::Tensor & wrapper_op_1(const at::Tensor & self) {
    // No device check
  // DeviceGuard omitted
  return at::native::op_1_kernel(self);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {
m.impl("op_1", TORCH_FN(wrapper_op_1));
}

namespace cpu {
at::Tensor & op_1(at::Tensor & self) {
  return wrapper_op_1(self);
}
} // namespace cpu
} // namespace at

namespace custom {

namespace {
at::Tensor & wrapper_op_2(const at::Tensor & self) {
    // No device check
  // DeviceGuard omitted
  return at::native::op_2_kernel(self);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {
m.impl("op_2", TORCH_FN(wrapper_op_2));
}

namespace cpu {
at::Tensor & op_2(at::Tensor & self) {
  return wrapper_op_2(self);
}
} // namespace cpu
} // namespace custom

```

The benefit for this change is that it unifies all the namespaces derived from custom ops. In the example above, there are:

1. `custom::native` for kernels
2. `custom::<dispatch_key>` e.g., `custom::cpu` for wrappers

This customized operator will have nothing to do with `at::native`, `at::cpu` etc.

Test Plan:
This is very hard to test. I will refactor this logic, abstract out some layers so it's testable. Will do it in coming PRs

**Static Docs Preview: executorch_cpp_api**
|[Full Site](https://our.intern.facebook.com/intern/staticdocs/eph/D37972772/V13/executorch_cpp_api/)|

|**Modified Pages**|

**Static Docs Preview: executorch_py_api**
|[Full Site](https://our.intern.facebook.com/intern/staticdocs/eph/D37972772/V13/executorch_py_api/)|

|**Modified Pages**|

Reviewed By: eddieluo01, bdhirsh

Differential Revision: D37972772

fbshipit-source-id: b59a05bb6c44b02a601adbaeda47a0a3ad0f0331
